Slack success/error  notification

parul-l opened the pull request:
Adding a slack success/error notification class:

* Created: `primrose.notifications.abstract_notification.py`:  
`AbstractNotification` class that enforces the post_message method

* Created: `primrose.notifications_utils`:  
Created `SlackClient` that inherits `AbstractNotification`. Other clients that are created that will be used for notifications should inherit the AbstractNotification class as well.

* Created: `src.notifications.success_notification.py`
`SuccessNotification` class inherits `AbstractSuccess` class. This is ideally the last node in the DAG. Given the parameters, this will instantiate a client (so far only `SlackClient` exists) and will post a DAG completion/custom message at the end of the DAG.

Created `get_notification_client` method that will instantiate a client given parameters. At the time being, we only have a slack client SlackClient.


* Modified: `primrose.dag_runner.py`. In the `.run` method, the code is modified as follows:
    * check if `metadata` exists and if so, check that there is a key called `notify_on_error` 
    * if `notify_on_error` exists, instantiate the client
    * in each `try/except` clause, the `except` clause posts an error message to the notification client

Sample config: (fyi `member_id` is optional - add it if you want to get pinged)
```
metadata:
    notify_on_error:
        client: SlackClient
        channel: your-slack-channel
        token: slack-api-token
        member_id: USomeUserID
implementation_config:
    reader_config:
        read_data:
            class: CsvReader
            filename: some/file/path
            destinations:
                - notification
    cleanup_config:
        notification:
            class: SuccessNotification
            client: SlackClient
            channel: your-slack-channel
            message: SUCCESS! DAG Completed!
            member_id: USomeUserID
            token: slack-api-token
```

    
Comments:
This is related to code structure. There is currently a `util.py` file in the `primrose` directory. I also create a `notification_util.py` file that in theory should be in a folder with the `util.py` file. Since I didn't want to change the structure around too much, I left it as is. But I think we should open an issue to just clean it up. Let me know your thoughts!

briangrahamww commented on primrose/dag_runner.py:
```diff
@@ -231,6 +233,22 @@ def run(self, dry_run=False):
+                raise(msg)
+        else:
+            client = None
+
```
I'm wondering if we would want the ability to allow the slack error notification parameters to optionally be set from environment variables too.  For our deployments, it would easy to point to slack notification credentials idb trigger file for our all of our batch jobs.  Also, I could set my local env to just notify myself over slack when I'm developing so as not to blast our main channels.

parul-l replied:
I like this idea @briangrahamww . Let me think about how best to integrate it.

briangrahamww replied:
@parul-l I believe we do the same thing in some nodes that we haven't migrated to primrose yet, like the redis writer/wiper

parul-l replied:
I will take a look. Thanks.

parul-l commented:
> One more thing. Documentation for the the new slack success node is automatically created from the docstring, but I would add to the readme the slack error notification feature. That built in notification capability is another feature users should be aware of.

Yes, you're right. I was thinking of adding something in the README (or maybe even a separate section).

parul-l commented:
Hey @briangrahamww. I think I have addressed your comments above. Here are the changes
* Corrected the bug that previously didn't allow for the notification node to be at any other point besides the last. (ie. excluded `destinations` in the parameters of the `get_notification_client` method.
* Allowed for parameters to be populated using environment variables
* Created a `README_NOTIFICATIONS.md` doc to explain how to use the feature. I'm not sure how to integrate this file in to the documentation, though.
Please take a look when you get the chance!

briangrahamww commented on README_NOTIFICATIONS.md:
````diff
@@ -0,0 +1,93 @@
+    token: slack-token
+    member_id: W012A3CDE
+    message: We did it!
+```
````
`W012A3CDE` is a dummy user_id, right? Just checking

parul-l replied:
yes! from the slack docs

briangrahamww commented on README_NOTIFICATIONS.md:
````diff
@@ -0,0 +1,93 @@
+      channel:
+      message: We did it!
+```
+The necessary keys to instantiate the client must be listed, and their values must be empty.
````
Could you make it so these don't have to be listed if we are using env vars instead?

parul-l replied:
Sure. This is done. I did it like this so that if you wanted both an error notification and a success notification, where for example, you wanted to be pinged on error (hence listing your `member_id` but didn't care for that in the success notification, then your default parameters would be those specified in the config, and any blanks would be filled in using the environment variables. In any case, perhaps no one would care on being pinged twice :).

parul-l replied:
@briangrahamww The current set up is that a notification client must be specified, and we will look for any environment variables associated to that client that are not already specified in the config file. That is, parameters specified in the config trump the environment variables.

Thinking about this some more, here are some things we can maybe do to allow more flexibility in specifying different parameters for say, the error notifications vs. node notifications (perhaps in another PR):

* In the config, specify the keys you want but have a default value for those keys for which you would like replaced by env variables.
Example
```
  cleanup_config:
    notification:
      class: ClientNotification
      client: SlackClient
      channel: my-channel
      token: USE-ENV-VAR
      message: We did it!
```
* Another option is to maybe only allow the token (and channel?) to be environment variables and everything else must be specified.

parul-l commented:
Thanks for all the reviews @briangrahamww! Do you mind approving the pull request so I can merge?
